### ISSUE TITLE
Error on out-of-line definition with constraints of template specialization

### ISSUE THREAD

**Issue by steve-downey** [llvm/llvm-project#109475]
Showed up during work on optional<T&> and making definitions out of line to produce the synopsis.

Reduced examplar:
```C++
#include <type_traits>

template <typename T>
struct X {};

template <class T>
struct X<T&> {
    template <typename U>
        requires(!(std::is_same_v<std::remove_cvref_t<U>, X>))
    X(U&& u);
};

template <typename T>
template <typename U>
    requires(!(std::is_same_v<std::remove_cvref_t<U>, X<T&>>))
X<T&>::X(U&& u) {}

```

https://compiler-explorer.com/z/v5Y14oTb8

The general converting constructor needs to be constrained out so as not to conflict with, or be a better match than, the copy constructor. 

<source>:16:8: error: out-of-line definition of 'X<type-parameter-0-0 &>' does not match any declaration in 'X<type-parameter-0-0 &>'
   16 | X<T&>::X(U&& u) {}
      |        ^
1 error generated.
Compiler returned: 1

I know there have been a number of  issues in this area, but kept finding ones that were fixed or seemed different.

**Comment by HerrCai0907** [comment]
The reason is in class, we inject class name and it cause subexpr in requires clause is different. Then the out-of-line definition don't treat as a redeclaration.
```
ElaboratedType 0x1378e1490 'X<T &>' sugar dependent
`-TemplateSpecializationType 0x1378e1440 'X<T &>' sugar dependent
  |-name: 'X' qualified
  | `-ClassTemplateDecl 0x1378c2160  X
  |-TemplateArgument type 'T &'
  | `-LValueReferenceType 0x1378e1010 'T &' dependent
  |   `-TemplateTypeParmType 0x1378e0f90 'T' dependent depth 0 index 0
  |     `-TemplateTypeParm 0x1378e0f38 'T'
  `-InjectedClassNameType 0x1378e11f0 'X<type-parameter-0-0 &>' dependent
    `-ClassTemplatePartialSpecialization 0x1378e10f0 'X'
```
```
ElaboratedType 0x1378e1ac0 'X<T &>' sugar dependent
`-TemplateSpecializationType 0x1378e1a70 'X<T &>' dependent
  |-name: 'X' qualified
  | `-ClassTemplateDecl 0x1378c2160  X
  `-TemplateArgument type 'T &'
    `-LValueReferenceType 0x1378e1a30 'T &' dependent
      `-TemplateTypeParmType 0x1378e1960 'T' dependent depth 0 index 0
        `-TemplateTypeParm 0x1378e1910 'T'

```

**Comment by zyn0217** [comment]
CC @sdkrystian in case you have fixed it somewhere in your patch.

**Comment by sdkrystian** [comment]
@zyn0217 I _think_ this should be fixed by #106585... unfortunately I'm not home so I can't verify at the moment

**Comment by zyn0217** [comment]
> @zyn0217 I _think_ this should be fixed by #106585... unfortunately I'm not home so I can't verify at the moment

I just updated my local build and can confirm that #106585 fixes this issue :)
